src/validr/_validator.pyx

```diff
@@ -115,6 +115,53 @@ def validator(bint string=False):
     return decorator
 
 
+_UNIQUE_CHECK_ERROR_MESSAGE = """
+Unable to check unique! only list of types below are available:
+    1. scalar type, eg: bool, int, float, str
+    2. list of scalar type
+    3. dict of scalar type value
+There are no general and fast unique check for non-hashable types,
+you should do unique check by yourself in this case.
+""".strip()
+
+
+def _is_scalar(validator):
+    return validator != 'dict' and validator != 'list'
+
+
+def _key_func_of_schema(schema):
+    if schema is None:
+        raise SchemaError(_UNIQUE_CHECK_ERROR_MESSAGE)
+
+    if schema.validator == 'dict':
+        if schema.items is None:
+            raise SchemaError(_UNIQUE_CHECK_ERROR_MESSAGE)
+        keys = []
+        for k, v in schema.items.items():
+            if not _is_scalar(v.validator):
+                raise SchemaError(_UNIQUE_CHECK_ERROR_MESSAGE)
+            keys.append(k)
+
+        def key_of(v):
+            return tuple(v[k] for k in keys)
```

Comment:
First of all, IIRC, you can't rely on the key order of the regular dict. At least, you have to sort them, but that's certanly not enough. You're relying on that keys for any dict instance in the list should be the same. But... are they?
What if some are optional. May be for now it won't be an error. But it's not something you should be keeping in your mind every time modifying your code not to make a regression. I strongly advice you to convert it to something like:
```python
return tuple((k,v) for k in sorted(keys))
```
You may think it's something trivial, but i recommend not to make a possible vulnerability.

Reply:
The `keys` is the same in validation process, it's determinated after schema compiled, and I'm not relying on it's order.
If some are optional, the value will be `None` or `''` (depends on inner validator).

Reply:
That may be the case, just my experience tells me to never make assumptions based on the default dict key ordering. In the end it explodes into your face. May be not always, but too often not to take it into account.